### DISCUSSION TITLE
overwrite default value of base class parameter within derived class

### DISCUSSION THREAD

**Discussion by arndflatten** [idaholab/moose#28743] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Some class `A` has an `addRequiredParam("pName","pDescr")` defined inside its `A::validParams()`.

My subclass B wants to make this `pName` an optional parameter and set some default value for it, but I dont see how, despite there must be a simple solution?

I tried
```
InputParameters
B::validParams()
{
  InputParameters params = A::validParams();
  params.makeParamNotRequired("pName");
  if(!params.isParamSetByUser("pName"))
    //* no such thing as setParam("pName", pValue) exists? So trying to add again:
    params.addParam("pName",newDefaultValue,"pDescr");
  return params;
}
```
but this doesn't take the newDefaultValue when i omit to define the pName in the input file. What am I missing out?

**Comment by GiudGiud** [comment, marked as answer]
in the validParams, you can re-add the parameter with a valid value

```
params.addParam<param_type (same as before!)>("pName", newDefaultvalue, "Description");
```
The if condition you have there should not be necessary. 

isParamSetByUser is not meant for use in validParams

**Reply by arndflatten** [reply]
Thank you. So, if i need to re-add the parameter again, also the  `params.makeParamNotRequired("pName")` is not necessary, right? 

And is there a way, how I can retrieve the `<param_type>` and `"oldDescr"` from the original baseClass-parameter automatically? 
Is the re-adding approach the considered-best way to achieve something like i want, or, is there an alternative way like using some `params.setDefaultValue("pName",pDefaultValue)` that I just overlooked?

Thanks again!

**Reply by GiudGiud** [reply]
I don't think there is a way to retrieve the parameter type.
You might be able to get the name by looping over the validParamsnof the base class
